x/observer/keeper/msg_server_add_observer.go

```diff
@@ -52,10 +52,15 @@ func (k msgServer) AddObserver(
 		return &types.MsgAddObserverResponse{}, nil
 	}
 
-	k.AddObserverToSet(ctx, msg.ObserverAddress)
+	// Add observer to the observer set
+	err = k.AddObserverToSet(ctx, msg.ObserverAddress)
+	if err != nil {
+		return &types.MsgAddObserverResponse{}, err
+	}
 	observerSet, _ := k.GetObserverSet(ctx)
 
 	k.SetLastObserverCount(ctx, &types.LastObserverCount{Count: observerSet.LenUint()})
+
 	EmitEventAddObserver(
```

Comment:
This looks these could be done in `AddObserverToSet`

Reply:
I agree with @lumtis but I lack some context:
- If the only way to add an observer is through `AddObserverToSet`, then emitting the event should be handled inside AddObserver, as the logic is intrinsic to it.
- If there are more ways of adding observers then we should refactor the event to be an optional callback to be passed to the N functions that add observers, such as `func SetObserver(foo, bar, eventToBeEmitted)`

The key thing here imho is to rely less and less with domain logic (If I invoke X, I have immediately to invoke Y).

Reply:
I agree with both of your points, however, I feel that emitting events in the main message is a pattern that we have been following across all the messages. And I would like to keep it the same for this message as well.

I did add the refactor the usage for `SetLastObserverCount` as it makes the the main message look cleaner 
83aed121f5b8beaaf55676da29d8b67f15f70c2f